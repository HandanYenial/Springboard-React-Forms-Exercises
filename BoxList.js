import React from "react";
import Box from "./Box";
import "./BoxList.css";
import { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import {v4 as uuid } from 'uuid';


//BoxList - Place your state that contains all of the boxes here.
//This component should render all of the Box components along with 
//the NewBoxForm component

const BoxList = () =>{
    const initialBox = [
        {
        id: uuid(),
        backgroundColor: "red",
        height: "100px",
        width: "100px"
        },
        {
        id: uuid(),
        backgroundColor: "blue",
        height: "100px",
        width: "100px"
        },
        {
        id: uuid(),
        backgroundColor: "green",
        height: "100px",
        width: "100px"
        }
    ];
    const [boxes, setBoxes] = useState(initialBox);
    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes,{...newBox , id:uuid()}]);
    }

    const remove = id =>{
        setBoxes(boxes => boxes.filter(box => box.id !== id));
    }
    
    /**const boxComponents = boxes.map(box => (
        <Box 
            key = {box.id}
            id = {box.id}
            width ={box.width}
            height ={box.height}
            handleRemove = {removeBox}
            color={box.color}
        />
    ))*/

    return(
        <div>
            <h1>Box List</h1>
            <NewBoxForm addBox={addBox}/>
            <div>
                {boxes.map(({ id, backgroundColor, height, width }) => 
                 <Box 
                     id ={id} 
                     backgroundColor={backgroundColor} 
                     width ={width} 
                     height ={height} 
                     key ={id}
                     handleRemove ={remove}

                    />)}
            </div>
        </div>
    )
}


export default BoxList;
