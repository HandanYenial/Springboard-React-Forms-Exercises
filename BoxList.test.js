import React from "react";
import {render, screen} from '@testing-library/react';
import  { fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

test("renders without crashing" , function(){
    render(<BoxList/>);
});

it ("matches snapshot" , function(){
    const {asFragment} = render(<BoxList/>);
    expect(asFragment()).toMatchSnapshot();
});

function addBox(boxList, height="4" , width="4", color="aqua"){
    const heightInput = boxList.screen.getByLabelText("Box Height");
    const widthInput  = boxList.screen.getByLabelText("Box Width");
    const colorInput  = boxList.screen.getByLabelText("Box Color");

    fireEvent.change(colorInput , { target : { value : color }});

    fireEvent.change(heightInput , { target : { value : height }});

    fireEvent.change(widthInput , { target : { value : width }});

    const button = boxList.screen.getByText("Add Box");

    fireEvent.click(button);

}

it ("can add a new box" , function(){
    const boxList = render(<BoxList />);

    // no boxes yet
    expect(boxList.screen.queryByText("Remove Box")).not.toBeInTheDocument();
    addBox(boxList);

    // one box now
    const removeButton = boxList.screen.getByText("Remove Box");
    expect(removeButton).toBeInTheDocument();
    expect(removeButton.previousSibling).toHaveStyle(`
          width : 4em;
          height: 4em;
          background-color: aqua;
          `);
    
    //expect form to be empty
    expect(boxList.screen.getAllByDisplayValue("")).toHaveLength(3);
});

it ("can remove a box" , function(){
    const boxList =render(<BoxList/>);
    addBox(boxList);

    const removeButton = boxList.getByText("Remove Box");
    fireEvent.click(removeButton);
    expect (removeButton).not.ToBeInTheDocument();

})