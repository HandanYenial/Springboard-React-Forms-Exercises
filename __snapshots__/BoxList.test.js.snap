// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Box List
    </h1>
    <form>
      <label
        for="backgroundColor"
      >
         Box Color 
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        placeholder="Box Color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
         Box Width :  
      </label>
      <input
        id="width"
        max="15"
        min="1"
        name="width"
        type="range"
        value=""
      />
      <label
        for="height"
      >
         Box Height: 
      </label>
      <input
        id="height"
        max="15"
        min="1"
        name="height"
        type="range"
        value=""
      />
      <button
        id="newBoxButton"
      >
         Add Box 
      </button>
    </form>
    <div>
      <div
        style="background-color: red;"
      >
        <button>
           X 
        </button>
      </div>
      <div
        style="background-color: blue;"
      >
        <button>
           X 
        </button>
      </div>
      <div
        style="background-color: green;"
      >
        <button>
           X 
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
